File: PC App/Julia/scan.jl
# scan.jl - --scan: the subnet-scanning client (Julia counterpart of
# PC App/App/registration.cpp and PC App/PyGame/registration.py). Queries every
# host on this machine's /24 for :<port>/robots and prints whatever answers.

import HTTP
import JSON3

function scan_host(ip::String, port::Int)
    try
        resp = HTTP.get("http://$ip:$port/robots"; readtimeout=1, connect_timeout=1, retry=false, status_exception=false)
        resp.status == 200 || return nothing
        # RIFT and NORA both serve {"authority": "...", "robots": [...]}, not a bare array.
        return JSON3.read(String(resp.body))
    catch
        return nothing
    end
end

function run_scan(local_ip::String, port::Int)
    base = join(split(local_ip, '.')[1:3], '.')    # assumes a /24
    println("Scanning $base.1-254 on port $port...")
    results = Vector{Any}(nothing, 254)
    sem = Base.Semaphore(50)
    @sync for i in 1:254
        @async begin
            Base.acquire(sem)
            try
                results[i] = scan_host("$base.$i", port)
            finally
                Base.release(sem)
            end
        end
    end
    println("\n=== Robots Found ===")
    for (i, body) in enumerate(results)
        (body === nothing || !haskey(body, :robots)) && continue
        for r in body.robots
            println("$(r.name) ($(r.type)) @ $base.$i\n  Capabilities: $(join(r.capabilities, " "))\n")
        end
    end
end

File: PC App/Go/scan.go
package main

// --scan: the subnet-scanning client (Go counterpart of PC App/App/registration.cpp
// and PC App/PyGame/registration.py). Queries every host on this machine's /24
// for :<port>/robots and prints whatever answers.

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"
)

type scanResult struct {
	IP     string
	Robots []Robot
}

func scanSubnet(base string, port int) []scanResult {
	client := &http.Client{Timeout: 500 * time.Millisecond}
	var (
		mu      sync.Mutex
		wg      sync.WaitGroup
		results []scanResult
		sem     = make(chan struct{}, 50)
	)
	for i := 1; i < 255; i++ {
		ip := fmt.Sprintf("%s.%d", base, i)
		wg.Add(1)
		sem <- struct{}{}
		go func() {
			defer wg.Done()
			defer func() { <-sem }()
			resp, err := client.Get(fmt.Sprintf("http://%s:%d/robots", ip, port))
			if err != nil {
				return
			}
			defer resp.Body.Close()
			// RIFT and NORA both serve {"authority": "...", "robots": [...]}, not a bare array.
			var body struct {
				Robots []Robot `json:"robots"`
			}
			if resp.StatusCode == http.StatusOK && json.NewDecoder(resp.Body).Decode(&body) == nil {
				mu.Lock()
				results = append(results, scanResult{IP: ip, Robots: body.Robots})
				mu.Unlock()
			}
		}()
	}
	wg.Wait()
	return results
}

func runScan(localIP string, port int) {
	base := localIP[:strings.LastIndex(localIP, ".")] // assumes a /24
	fmt.Printf("Scanning %s.1-254 on port %d...\n", base, port)
	results := scanSubnet(base, port)
	fmt.Println("\n=== Robots Found ===")
	for _, res := range results {
		for _, r := range res.Robots {
			fmt.Printf("%s (%s) @ %s\n  Capabilities: %s\n\n", r.Name, r.Type, res.IP, strings.Join(r.Capabilities, " "))
		}
	}
}
